### PR TITLE
feat(Auth): Adding TOTP state machine actions

### PR DESCRIPTION
**Pull request by harsh62**
### TOTP PR Number # 5

Depends on: 
* #3040 
* #3041 
* #3042
* #3043 

Changes
* Adds implementation of actions needed by the State Machine
* Update UserPoolSignIn Helper to add details about TOTP

_NOTE: The unit and integration tests will fail until all changes make it to the feature branch_
(The PR is directed towards a feature branch)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/SignIn/SoftwareTokenSetup/CompleteTOTPSetup.swift [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+            let responseEvent = try await UserPoolSignInHelper.sendRespondToAuth(
+                request: input,
+                for: username,
+                signInMethod: signInEventData.signInMethod,
+                environment: userpoolEnv)
+            logVerbose("\(#fileID) Sending event \(responseEvent)",
+                       environment: environment)
+            await dispatcher.send(responseEvent)
+            // TODO: HS:
```
Comment (ruisebas) [inline AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/SignIn/SoftwareTokenSetup/CompleteTOTPSetup.swift:84]: Remember to remove commented out code before releasing

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Support/Helpers/UserPoolSignInHelper.swift [hunk 2]
```diff
@@ -54,12 +63,18 @@ struct UserPoolSignInHelper: DefaultLogger {
...
         case .customChallenge:
             return .init(nextStep: .confirmSignInWithCustomChallenge(challenge.parameters))
         case .newPasswordRequired:
             return .init(nextStep: .confirmSignInWithNewPassword(challenge.parameters))
-        case .unknown:
-            throw AuthError.unknown("Challenge not supported", nil)
+        case .selectMFAType:
+            return .init(nextStep: .continueSignInWithMFASelection(challenge.getAllowedMFATypesForSelection))
+        case .setUpMFA:
+            throw AuthError.unknown("Invalid state flow. setUpMFA is handled internally in `SignInState.resolvingTOTPSetup` state.")
```
Comment (royjit) [inline AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Support/Helpers/UserPoolSignInHelper.swift:75]: Should we use `AuthError.invalidState` here?

**Review by thisisabhash — APPROVED** [review]

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/SignIn/SoftwareTokenSetup/CompleteTOTPSetup.swift (hunk 1 above)
Comment (harsh62) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Actions/SignIn/SoftwareTokenSetup/CompleteTOTPSetup.swift:84]: Yep.. This is something I still need to work on.. Will remove it in a follow up PR.

File: AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Support/Helpers/UserPoolSignInHelper.swift (hunk 2 above)
Comment (harsh62) [reply AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Support/Helpers/UserPoolSignInHelper.swift:75]: Nice catch.. Fixed.

**Review by royjit — APPROVED** [review]

**Pull request merged**